pkg/apis/experimental/v1/types.go

```diff
@@ -462,3 +462,79 @@ type JobCondition struct {
 	// Human readable message indicating details about last transition.
 	Message string `json:"message,omitempty"`
 }
+
+// Ingress encapsulates the inputs needed to config the load-balancer.
+// It represents a rule collection of inbound connections from the external network
+// that would be satisfied by a load-balancer.
+type Ingress struct {
+	v1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	v1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the desired state of the Ingress.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
+	Spec IngressSpec `json:"spec,omitempty"`
+
+	// Status is the current state of the Ingress.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
+	Status IngressStatus `json:"status,omitempty"`
+}
+
+// IngressList is a collection of Ingress.
+type IngressList struct {
+	v1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	v1.ListMeta `json:"metadata,omitempty"`
+
+	// Items is the list of ingress.
+	Items []Ingress `json:"items"`
+}
+
+// IngressSpec describes the ingress the user wishes to exist.
+type IngressSpec struct {
+	// PathMap defines a list of rules which mapping multiple hosts and paths to the backend services.
+	// The rules will be used to configure the corresponding load-balancer.
+	// Example: Host + Path -> Service Backend (Service Name + Service Protocol + Service Port)
+	PathMap []IngressMap `json:"pathMap,omitempty"`
```

Comment:
Technically it's http plus TLS, tls can front anything.  So can we break
the explicit tie to http and make it part of the ingress spec (host is
generic to tls+sni, http, a few others, path is not)

On Sep 17, 2015, at 6:55 PM, Brian Grant <notifications@github.com> wrote:

In pkg/apis/experimental/v1/types.go
<https://github.com/kubernetes/kubernetes/pull/13947#discussion_r39808831>:

> +type IngressList struct {
> +	v1.TypeMeta `json:",inline"`
> +	// Standard object's metadata.
> +	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
> +	v1.ListMeta `json:"metadata,omitempty"`
> +
> +	// Items is the list of ingress.
> +	Items []Ingress `json:"items"`
> +}
> +
> +// IngressSpec describes the ingress the user wishes to exist.
> +type IngressSpec struct {
> +	// PathMap defines a list of rules which mapping multiple hosts and paths to the backend services.
> +	// The rules will be used to configure the corresponding load-balancer.
> +	// Example: Host + Path -> Service Backend (Service Name + Service Protocol + Service Port)
> +	PathMap []IngressMap `json:"pathMap,omitempty"`

@smarterclayton <https://github.com/smarterclayton> mentioned other future
types of ingress:
https://github.com/kubernetes/kubernetes/pull/14095/files#r39790321

This resource doesn't seem very extensible to these other types. The
structure is very specific to host+path -> target mapping, as opposed to
generic-something -> target mapping. That suggests to me that we'll need
new ingress API types for those other cases, in which case we should make
this API type's name more specific, like HTTPIngress.

@thockin <https://github.com/thockin> WDYT?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13947/files#r39808831>.